### PR TITLE
SNO+: XL3 review

### PR DESCRIPTION
**Pull request by BenLand100**
Cleaned up and commented code.
Removed all MEGABUNDLE related code.
Appropriately number XL3 packets at low level.
Use @finally appropriately for locks and @try blocks.

Tested on teststand. Pending testing on detector.

### CODE REVIEW

File: Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m [hunk 1]
```diff
@@ -352,17 +301,23 @@ - (void) setCrateName:(NSString*)aCrateName
...
-		//[[NSNotificationCenter defaultCenter] postNotificationName:XL3_LinkIPNumberChanged object:self];
+		crateName = [aCrateName copy];    	
 	}
 }	
 
+/**
+ * Returns an NSArray of 17 NSNumber objects representing the FIFO status
+ * of the FECs in the this XL3's crate or nil if unknown.
+ */
 - (NSArray*) fifoStatus
```
Comment (tlatorre-uchicago) [inline Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m]: can be deleted.

File: Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m [hunk 2]
```diff
@@ -352,17 +301,23 @@ - (void) setCrateName:(NSString*)aCrateName
...
 {
     return fifoStatus; //may return nil
 }
 
+/**
+ * Stores an NSArray of 17 NSNumber objects representing the FIFO status
+ * of the FECs in the this XL3's crate.
+ */
 - (void) setFifoStatus:(NSArray *)aFifoStatus
```
Comment (tlatorre-uchicago) [inline Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m]: same here.

File: Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m [hunk 3]
```diff
@@ -372,6 +327,10 @@ - (void) setFifoStatus:(NSArray *)aFifoStatus
     _fifoTimeStamp = [[NSDate alloc] init];
 }
 
+/**
+ * Converts the raw memory form of fifo status to a form appropriate for
+ * setFifoStatus and calls setFifoStatus.
+ */
 - (void) copyFifoStatus:(int32_t*)aStatus
```
Comment (tlatorre-uchicago) [inline Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m]: and here.

File: Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m [hunk 4]
```diff
@@ -395,41 +354,55 @@ - (void) copyFifoStatus:(int32_t*)aStatus
     [self setReadFifoFlag:YES];
 }
 
+/**
+ * Returns a pointer to the raw memory form of the result of fifoStatus.
+ */
 - (unsigned long*) fifoBundle
```
Comment (tlatorre-uchicago) [inline Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m]: and here.

File: Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m [hunk 5]
```diff
@@ -395,41 +354,55 @@ - (void) copyFifoStatus:(int32_t*)aStatus
...
 	MultiCommand* theMultiCommand = (MultiCommand*) aMultiCmdPacket.payload;
 	if (needToSwap) {
 		unsigned int i = 0;
 		for (i = 0; i < theMultiCommand->howMany; i++) {
 			Command* command = &(theMultiCommand->cmd[i]);
-			command->cmdNum = swapLong(command->packetNum);
+			command->cmdNum = swapLong(command->cmdNum);
 			command->packetNum = swapShort(command->packetNum);
 			command->address = swapLong(command->address);
 			command->data = swapLong(command->data);
```
Comment (tlatorre-uchicago) [inline Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m]: would vote for htonl here. The XL3 is guaranteed to be big endian.

File: Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m [hunk 6]
```diff
@@ -442,17 +415,15 @@ - (XL3Packet*) executeMultiCmd
 		[self sendXL3Packet:&aMultiCmdPacket];
 	}
 	@catch (NSException* localException) {
-		NSLog(@"%@ MultiCmd failed.\n", [self crateName]);
-		//@throw localException;
+		NSLogColor([NSColor redColor],@"%@ MultiCmd failed.\n", [self crateName]);
```
Comment (tlatorre-uchicago) [inline Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m]: If this fails, how does the caller know? Or is this only used for commands from the GUI? If so, that should be specified in the comment at the top.

File: Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m [hunk 7]
```diff
@@ -442,17 +415,15 @@ - (XL3Packet*) executeMultiCmd
...
-		aMultiCmdPacket.header.packetType = swapShort(aMultiCmdPacket.header.packetType);
 		theMultiCommand->howMany = swapLong(theMultiCommand->howMany);
 		unsigned int i = 0;
 		for (i = 0; i < theMultiCommand->howMany; i++) {
 			Command* command = &(theMultiCommand->cmd[i]);
-			command->cmdNum = swapLong(command->packetNum);
+			command->cmdNum = swapLong(command->cmdNum);
 			command->packetNum = swapShort(command->packetNum);
 			command->address = swapLong(command->address);
 			command->data = swapLong(command->data);
```
Comment (tlatorre-uchicago) [inline Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m]: ntohl/ntohs

File: Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m [hunk 8]
```diff
@@ -476,48 +452,58 @@ - (BOOL) multiCmdFailed
...
+    xl3Packet.header.numBundles = 0;
+    memcpy(xl3Packet.payload, payload, XL3_PAYLOAD_SIZE);
     
+    [commandSocketLock lock]; 
     @try {
-        [commandSocketLock lock]; //begin critical section
-        [self writePacket:(char*) &aPacket];
-        [commandSocketLock unlock]; //end critical section
+        [self writePacket:&xl3Packet];
+        packetNum = needToSwap ? swapShort(xl3Packet.header.packetNum) : xl3Packet.header.packetNum;
```
Comment (tlatorre-uchicago) [inline Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m]: ntohs

File: Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m [hunk 9]
```diff
@@ -526,31 +512,39 @@ - (void) sendCommand:(uint8_t) aCmd withPayload:(char *) payload expectResponse:
...
+ * and the result is coppied into the same variable. Throws exceptions on failure.
+ */
 - (void) sendCommand:(uint8_t) aCmd toAddress:(uint32_t) address withData:(uint32_t *) value
 {
 	char payload[XL3_PAYLOAD_SIZE];
 	Command* command = (Command*) payload;
 		
-	command->cmdNum = aCmd;
-	command->packetNum = 0; // todo: figure out what are these two good for...
+	command->cmdNum = 0;
```
Comment (BenLand100) [inline Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m]: Nope, unused by the XL3 - pretty sure this was put here originally just to put something here. 
aCmd is unused in this method unfortunately - will update comment to reflect that.

**Comment by BenLand100** [comment]
For the packet parsing the main change was switching from an if-else cascade to a switch statement and re indenting to match - the actual parsing hasn't changed. However, if we know a way to generate SCREWED and ERROR packets on the test stand I'll double check.

**Comment by tlatorre-uchicago** [comment]
The screwed can definitely be reproduced by sending pedestals to a card at > 5 kHz or so. I'm not sure about the error packet, I'll look into it.

File: Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m (hunk 6 above)
Comment (BenLand100) [inline Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m]: I believe it is used by more places than the GUI however the comment reads - "Does not raise exception on failure. Check multiCmdFailed for success or inspect the result packet."

File: Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m [hunk 10]
```diff
@@ -398,36 +330,31 @@ - (void) addMultiCmdToAddress:(long)anAddress withValue:(long)aValue
...
-		for (i = 0; i < theMultiCommand->howMany; i++) {
-			Command* command = &(theMultiCommand->cmd[i]);
-			command->cmdNum = swapLong(command->cmdNum);
-			command->packetNum = swapShort(command->packetNum);
-			command->address = swapLong(command->address);
-			command->data = swapLong(command->data);
-		}
+	
+	theMultiCommand->howMany = ntohl(theMultiCommand->howMany);
+	unsigned int i = 0;
```
Comment (tlatorre-uchicago) [inline Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m]: double definition of i

Comment (BenLand100) [inline Source/Objects/Custom Hardware/SNO+/XL3/XL3_Link.m]: Ah yeah should note I haven't actually checked that this compiles yet - was going to work on that tonight.

**Comment by BenLand100** [comment]
@tlatorre-uchicago double checking things on the teststand I encountered a possible issue with the xl3 server - looks like if it's unable to post an alarm to the DB (presumably crate 19 alarms do not exist) it also doesn't send the screwed packet to ORCA (I checked with master - nothing printed to the log there either). See the following error:
```
21 Jan 03:07:41  xl3 # post_alarm function failed: ERROR:  insert or update on table "alarms" violates foreign key constraint "alarms_alarm_id_fkey"
    DETAIL:  Key (alarm_id)=(10019) is not present in table "alarm_descriptions".
    CONTEXT:  SQL statement "INSERT INTO alarms (id, alarm_id) VALUES (_id, $1)"
    PL/pgSQL function post_alarm(bigint) line 10 at SQL statement
21 Jan 03:07:43  xl3 # xl3 19: screwed. FEC FIFO overflowed. data lost for fecs: 8.
```

**Comment by tlatorre-uchicago** [comment]
Yeah, I forgot that the XL3 server actually never sends screwed packets back to clients because it's not obvious who to send it to if there are multiple clients connected.

**Comment by tlatorre-uchicago** [comment]
merged.

**Pull request closed without merging**